### PR TITLE
[#27] [Backend] As a logged-in user, I can submit my answers to the survey questions

### LINKED ISSUE #27: [Backend] As a logged-in user, I can submit my answers to the survey questions
## Why
The user can submit their answers to the survey questions.

## Acceptance Criteria
- The API endpoint `/api/v1/responses` is called with the following body:
  ```
  {
    "survey_id": "{{survey_id}}",
    "questions": [
        {
            "id": "{{question_id}}",
            "answers": [
                {
                    "id": "{{answer_id}}"
                }
            ]
        },
        {
            "id": "{{question_id}}",
            "answers": [
                {
                    "id": "{{answer_id}}",
                    "answer": "My answer"
                }
            ]
        }
    ]
  }
  ```
- Unit tests are added and passed.

## Resources
The API documentation is available on this publicly accessible [Postman collection](https://documenter.getpostman.com/view/11835486/UVsPNPfU).

### PR DESCRIPTION
**Pull request by kaungkhantsoe**
- Closes #27 

## What happened 👀

The user can submit their answers to the survey questions.

## Insight 📝

- Add logic in repository
- Add usecase
- Add unit tests

## Proof Of Work 📹

Screenshot 2023-05-19 at 11 48 51 AM (2)
Screenshot 2023-05-19 at 11 45 24 AM (2)

### CODE REVIEW

File: lib/api/api_service.dart [hunk 1]
```diff
@@ -33,4 +34,7 @@ abstract class ApiService {
   Future<SurveyDetailDataResponse> getSurveyDetail(
     @Path('surveyId') String surveyId,
   );
+
+  @POST('api/v1/responses')
+  Future<void> submitSurveys(@Body() SubmitSurveysRequest submitSurveysRequest);
```
Comment (manh-t) [inline lib/api/api_service.dart:39]:
```suggestion
  Future<void> submitSurvey(@Body() SubmitSurveyRequest submitSurveyRequest);
```

File: lib/api/repository/survey_repository.dart [hunk 2]
```diff
@@ -29,6 +30,8 @@ abstract class SurveyRepository {
   });
 
   Future<SurveyDetailDataResponse> getSurveyDetail(String surveyId);
+
+  Future<void> submitSurveys(SubmitSurveysRequest submitSurveyRequest);
```
Comment (manh-t) [inline lib/api/repository/survey_repository.dart:34]:
```suggestion
  Future<void> submitSurvey(SubmitSurveyRequest submitSurveyRequest);
```

File: lib/model/request/submit_surveys_request.dart [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:survey_flutter/model/request/survey_question_request.dart';
+
+part 'submit_surveys_request.g.dart';
+
+@JsonSerializable()
+class SubmitSurveysRequest {
```
Comment (manh-t) [inline lib/model/request/submit_surveys_request.dart:7]:
```suggestion
class SubmitSurveyRequest {
```

**Review by manh-t — COMMENTED** [review]
Because we are submitting the questions for a survey, it should be singular `submitSurvey` 😄

File: lib/api/api_service.dart [hunk 4]
```diff
@@ -33,4 +34,7 @@ abstract class ApiService {
   Future<SurveyDetailDataResponse> getSurveyDetail(
     @Path('surveyId') String surveyId,
   );
+
+  @POST('api/v1/responses')
+  Future<void> submitSurvey(@Body() SubmitSurveyRequest submitSurveysRequest);
```
Comment (manh-t) [inline lib/api/api_service.dart:39]:
```suggestion
  Future<void> submitSurvey(@Body() SubmitSurveyRequest submitSurveyRequest);
```

File: lib/model/request/submit_surveys_request.dart
Comment (manh-t) [inline lib/model/request/submit_surveys_request.dart:1]: @kaungkhantsoe please rename the file as well

File: lib/model/request/submit_surveys_request.dart [hunk 5]
```diff
@@ -0,0 +1,22 @@
+import 'package:freezed_annotation/freezed_annotation.dart';
+import 'package:survey_flutter/model/request/survey_question_request.dart';
+
+part 'submit_surveys_request.g.dart';
```
Comment (manh-t) [inline lib/model/request/submit_surveys_request.dart:4]:
```suggestion
part 'submit_survey_request.g.dart';
```

File: lib/usecases/submit_surveys_use_case.dart
Comment (manh-t) [inline lib/usecases/submit_surveys_use_case.dart:1]: please rename the file to `submit_survey_use_case.dart`

File: lib/usecases/submit_surveys_use_case.dart [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+import 'package:survey_flutter/usecases/base/base_use_case.dart';
+
+final submitSurveysUseCaseProvider = Provider((ref) {
+  return SubmitSurveysUseCase(ref.watch(surveyRepositoryProvider));
+});
+
+class SubmitSurveysUseCase extends UseCase<void, SubmitSurveyRequest> {
+  final SurveyRepository _surveyRepository;
+
+  SubmitSurveysUseCase(this._surveyRepository);
```
Comment (manh-t) [inline lib/usecases/submit_surveys_use_case.dart:13]: update the class's name too

File: test/usecases/submit_surveys_use_case_test.dart [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      expect(result, isA<Success>());
+    });
+
+    test(
+      'When calling submitSurveys with SubmitSurveysRequest unsuccessfully, it emits failed',
+      () async {
+        when(mockSurveyRepository.submitSurvey(any)).thenThrow(MockDioError());
+
+        final result =
+            await submitSurveysUseCase.call(MockUtil.submitSurveysRequest);
```
Comment (manh-t) [inline test/usecases/submit_surveys_use_case_test.dart:35]:
```suggestion
            await submitSurveysUseCase.call(MockUtil.submitSurveyRequest);
```

File: test/mocks/mock_util.dart [hunk 8]
```diff
@@ -59,4 +62,17 @@ class MockUtil {
...
+
+  static SurveyQuestionRequest surveyQuestionRequest = SurveyQuestionRequest(
+    id: 'id',
+    answers: List.empty(),
+  );
+
+  static SurveyAnswerRequest surveyAnswerRequest =
+      SurveyAnswerRequest(id: 'id');
+
+  static SubmitSurveyRequest submitSurveysRequest = SubmitSurveyRequest(
```
Comment (manh-t) [inline test/mocks/mock_util.dart:74]:
```suggestion
  static SubmitSurveyRequest submitSurveyRequest = SubmitSurveyRequest(
```

File: test/usecases/submit_surveys_use_case_test.dart [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+  group('SubmitSurveysUseCase', () {
+    MockSurveyRepository mockSurveyRepository = MockSurveyRepository();
+
+    late SubmitSurveysUseCase submitSurveysUseCase;
+
+    setUp(() =>
+        submitSurveysUseCase = SubmitSurveysUseCase(mockSurveyRepository));
+
+    test(
+        'When calling submitSurveys with SubmitSurveysRequest successfully, it emits success',
```
Comment (manh-t) [inline test/usecases/submit_surveys_use_case_test.dart:19]:
```suggestion
        'When calling submitSurvey with SubmitSurveyRequest successfully, it emits success',
```

File: test/usecases/submit_surveys_use_case_test.dart [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+      when(mockSurveyRepository.submitSurvey(any))
+          .thenAnswer((_) => Future.value());
+
+      final result =
+          await submitSurveysUseCase.call(MockUtil.submitSurveysRequest);
+      expect(result, isA<Success>());
+    });
+
+    test(
+      'When calling submitSurveys with SubmitSurveysRequest unsuccessfully, it emits failed',
```
Comment (manh-t) [inline test/usecases/submit_surveys_use_case_test.dart:30]:
```suggestion
      'When calling submitSurvey with SubmitSurveyRequest unsuccessfully, it emits failed',
```

File: test/usecases/submit_surveys_use_case_test.dart [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import '../mocks/mock_util.dart';
+
+void main() {
+  group('SubmitSurveysUseCase', () {
+    MockSurveyRepository mockSurveyRepository = MockSurveyRepository();
+
+    late SubmitSurveysUseCase submitSurveysUseCase;
+
+    setUp(() =>
+        submitSurveysUseCase = SubmitSurveysUseCase(mockSurveyRepository));
```
Comment (manh-t) [inline test/usecases/submit_surveys_use_case_test.dart:16]:
```suggestion
        submitSurveyUseCase = SubmitSurveyUseCase(mockSurveyRepository));
```

**Review by manh-t — CHANGES_REQUESTED** [review]
there are still some places which are having `s` in `SubmitSurveys`. Please update them 🙏

**Review by manh-t — APPROVED** [review]

**Review by ducbm051291 — APPROVED** [review]

**Pull request merged**